docs: :memo: revised section on properties to be only about the class

lwjohnst86 opened the pull request:
## Description

If we have this class, it will replace all these other functions. I haven't written much in them since they would basically be a 1-to-1 relationship with the spec itself.

## Reviewer Focus

<!-- Select quick/in-depth as necessary -->
This PR needs a quick review.

lwjohnst86 committed:
docs: :memo: revised section on properties to be only about the class

lwjohnst86 commented on docs/design/implementation/python-functions.qmd:
```diff
@@ -409,35 +409,21 @@ also creates the necessary directory if it doesn't exist.
 
 ::: {.callout-note collapse="true"}
-### `view_properties()`
+### `Properties(...)`
```
The problem is you can't nest dataclasses, at least not that I could see from the documentation. You can set a field in one dataclass as another dataclass, but the dataclass is created as a top level class, not a subclass of another dataclass. We could write our own code to do that, but it looks complicated and I don't see the value of that.

signekb replied:
Ah, that is actually also what I meant. Like 
```
class Package: 
	...
	resources: list[Resource] = field(default_factory=list)
	...

@dataclass
class Resource: 
	title: str 
	... 
```

I'm not sure I understand why we would need the `Properties` level still?

signekb replied:
So `Properties` would look something like this? ☺️ 
```
class Properties: 
	Package: Package
	Resource: Resource
	...
```

martonvago replied:
Sorry, just trying to catch up with this:
- You can nest dataclasses as far as I can see. I think you can define a class basically anywhere you want in Python.
- Nested classes are not subclasses of the outer class. If we wanted subclasses we would use inheritance. Inheritance would e.g. allow us to share logic between classes.
- I think the two contending ways of organising the classes are currently:
1. Separate classes:
```python
@dataclass
class SourceProperties:
    title: str = ""
    ...

# similar for ContributorProperties, LicenseProperties

@dataclass
class ResourceProperties:
    sources: list[SourceProperties] = field(default_factory=list)
    ...

@dataclass
class PackageProperties:
    resources: list[ResourceProperties] = field(default_factory=list)
    ...
```
I think this is fairly clear.

2. Nested classes:
```python
@dataclass
class Properties:

    @dataclass
    class Source:
        title: str = ""
        ...

    # similar for Contributor, License

    @dataclass
    class Resource:
        sources: list["Properties.Source"] = field(default_factory=list)
        ...

    @dataclass
    class Package:
        resources: list["Properties.Resource"] = field(default_factory=list)
        ...
```
Here, the nesting is purely organisational, as `Properties` does not really correspond to anything by itself, does not represent a "whole" that consists of "parts" -- which is usually what class nesting expresses in other languages, I think. I initially felt a bit uneasy about whether this was a semantically correct use of class nesting, but now I think it's fine to do if it helps us ([relevant, but not decisive discussion](https://stackoverflow.com/questions/3576596/is-it-a-good-idea-to-using-class-as-a-namespace-in-python)).

lwjohnst86 committed:
docs: :memo: small edits to clarify the Properties object

lwjohnst86 committed:
docs: :memo: split out dataclasses and include example usage

lwjohnst86 commented:
@martonvago and @signekb I've taken the feedback and split the properties into individual dataclasses as well as added some example usage, just so it is a bit clearer what they would do/be used for.

martonvago reviewed (changes requested):
Sorry for the novel 😅

lwjohnst86 committed:
chore: :fire: don't need these anymore (#694)

## Description

This is a stacked PR on top of the changes to the description of the
properties. This removes the property functions from the init file.

## Reviewer Focus

<!-- Select quick/in-depth as necessary -->
This PR needs a quick review.

signekb commented on docs/design/implementation/python-functions.qmd:
````diff
@@ -406,38 +406,120 @@ also creates the necessary directory if it doesn't exist.
+```
+
+```
+LicenseProperties(...)
````
```suggestion
LicenseProperties(name=None, path=None, title=None)
```

signekb commented on docs/design/implementation/python-functions.qmd:
````diff
@@ -406,38 +406,120 @@ also creates the necessary directory if it doesn't exist.
-TODO: This function is still being developed. Used to help create a
-properties object to pass around.
+```
+TableSchemaProperties(...)
````
```suggestion
TableSchemaProperties(fields=[], missingValues=[], primaryKey=[], foreignKeys=[])
```
Or something like this? ☺️

lwjohnst86 committed:
Apply suggestions from code review

Co-authored-by: martonvago <57952344+martonvago@users.noreply.github.com>
Co-authored-by: Signe Kirk Brødbæk <40836345+signekb@users.noreply.github.com>

lwjohnst86 committed:
docs: :pencil2: small fix related to properties all being classes

lwjohnst86 committed:
Merge branch 'main' into docs/describe-properties-class
